lib/puma/request.rb

```diff
@@ -0,0 +1,438 @@
+# frozen_string_literal: true
+
+module Puma
+
```

Comment:
It looks like you've found a seam but haven't quite cut through all the way yet. I think you can go all the way to a composed object:

```ruby
while true 
  case Request.new(client,buffer).handle
```

... and then in `lib/puma/server/request.rb`, everything except `handle_request` can be marked private. 

This would greatly reduce the API surface of Server. Right now this moves methods around in different files, but with a composed object we could really make Server a lot smaller.

```

Reply:
I did think about this.  Two items:

1. It would be a breaking API change.  `handle_request` and `normalize_env` are currently public methods of Server.

2. Coupling - the following Server instance variables are used in Request code:
```
@app              available
@early_hints      available
@events           available
@queue_requests   hidden
@thread_pool      hidden
```

We could add a couple of readers and pass the server instance into the class init?  Also, re class init, I'm not sure if the buffer should/can be contained within Request, rather than be passed in.

I thought that the current layout would maintain the API, but make it easy to create a class when v6.0.0 was being considered.

But, re API changes, the earlier issue with the API changes I made affected parts of the API that could be used for Puma startup/setup.  These changes are more 'runtime'.  Possibly less likely to cause issues...

Reply:
OK, got it. I'll re-review this with a mind to "this is a temporary stopping point to an extraction"